go-controller/pkg/util/ndp/ra.go

```diff
@@ -0,0 +1,101 @@
+	"github.com/google/gopacket/layers"
+	"github.com/mdlayher/socket"
+	"golang.org/x/sys/unix"
+)
+
+// RouterAdvertisement with mac, ips and lifetime field to send
+type RouterAdvertisement struct {
+	SourceMAC, DestinationMAC net.HardwareAddr
+	SourceIP, DestinationIP   net.IP
+	Lifetime                  uint16
+}
+
+// SendRouterAdvertisements will send variadic RAs.
+// Requires raw socket capabilities, since the RAs src MAC
+// and IP are not the ones on the LRP.
+func SendRouterAdvertisements(interfaceName string, ras ...RouterAdvertisement) error {
+	iface, err := net.InterfaceByName(interfaceName)
+	if err != nil {
+		return err
+	}
+	c, err := socket.Socket(syscall.AF_PACKET, syscall.SOCK_RAW, syscall.ETH_P_ALL, "ra", nil)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+	serializedRAs := [][]byte{}
+	for _, ra := range ras {
+		serializeBuffer := gopacket.NewSerializeBuffer()
+		ethernetLayer := layers.Ethernet{
+			DstMAC:       ra.DestinationMAC,
+			SrcMAC:       ra.SourceMAC,
+			EthernetType: layers.EthernetTypeIPv6,
+		}
+		ip6Layer := layers.IPv6{
+			Version:    6,
+			NextHeader: layers.IPProtocolICMPv6,
+			HopLimit:   255,
+			SrcIP:      ra.SourceIP,
+			DstIP:      ra.DestinationIP,
+		}
+		icmp6Layer := layers.ICMPv6{
+			TypeCode: layers.CreateICMPv6TypeCode(layers.ICMPv6TypeRouterAdvertisement, 0),
+		}
+		if err := icmp6Layer.SetNetworkLayerForChecksum(&ip6Layer); err != nil {
+			return err
+		}
+
+		// https://datatracker.ietf.org/doc/html/rfc4861
+		// 1-bit "Managed address configuration" flag.  When
+		// set, it indicates that addresses are available via
+		// Dynamic Host Configuration Protocol [DHCPv6].
+		managedAddressFlag := uint8(0x80)
+
+		//  https://datatracker.ietf.org/doc/html/rfc4191#section-2.2
+		//	Prf (Default Router Preference):
+		//		2-bit signed integer.  Indicates whether to prefer this
+		//		router over other default routers.  If the Router Lifetime
+		//		is zero, the preference value MUST be set to (00) by the
+		//		sender and MUST be ignored by the receiver.  If the Reserved
+		//		(10) value is received, the receiver MUST treat the value as
+		//		if it were (00).
+		defaultRoutePreferenceFlag := uint8(0x08)
+		if ra.Lifetime == 0 {
+			defaultRoutePreferenceFlag = uint8(0x00)
+		}
+
+		raLayer := layers.ICMPv6RouterAdvertisement{
+			HopLimit:       255,
+			Flags:          managedAddressFlag | defaultRoutePreferenceFlag,
+			RouterLifetime: ra.Lifetime,
+			ReachableTime:  0,
+			RetransTimer:   0,
+			Options: layers.ICMPv6Options{{
+				Type: layers.ICMPv6OptSourceAddress,
+				Data: ra.SourceMAC,
+			}},
+		}
+		if err := gopacket.SerializeLayers(serializeBuffer, gopacket.SerializeOptions{ComputeChecksums: true, FixLengths: true},
+			&ethernetLayer,
+			&ip6Layer,
+			&icmp6Layer,
+			&raLayer,
+		); err != nil {
+			return err
+		}
+		serializedRAs = append(serializedRAs, serializeBuffer.Bytes())
+	}
+	for _, serializedRA := range serializedRAs {
```

Comment:
This file I heavily trust your skills haha
I wish we had a library help for sending out these RAs but ..

Reply:
well now we have.